### ISSUE TITLE
Incorrect trait-object warning when returning a type-alias-impl-trait

### ISSUE THREAD

**Issue by Aaron1011** [rust-lang/rust#69147]
The following code:

```rust
#![feature(type_alias_impl_trait)]

type Foo = impl Copy;
fn bar(val: &u8) -> Foo + 'static {
    true
}
```

produces the following messages:

```
error[E0404]: expected trait, found opaque type `Foo`
 --> src/lib.rs:4:21
  |
4 | fn bar(val: &u8) -> Foo + 'static {
  |                     ^^^ not a trait

warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:4:21
  |
4 | fn bar(val: &u8) -> Foo + 'static {
  |                     ^^^^^^^^^^^^^ help: use `dyn`: `dyn Foo + 'static`
  |
  = note: `#[warn(bare_trait_objects)]` on by default
```

The error message is correct - `Foo` is not a trait, so writing `+ 'static` makes no sense. However, the warning message is completely inapplicable - `Foo` is not a trait, so adding `dyn` will just produce a different error.

We should suppress this warning if the user didn't actually write down the name of a trait.

**Comment by Centril** [comment]
And without the feature gate:
```rust
fn bar() -> u8 + 'static {
    0
}
```
==>
```rust

error[E0404]: expected trait, found builtin type `u8`
 --> src/lib.rs:1:13
  |
1 | fn bar() -> u8 + 'static {
  |             ^^ not a trait

warning: trait objects without an explicit `dyn` are deprecated
 --> src/lib.rs:1:13
  |
1 | fn bar() -> u8 + 'static {
  |             ^^^^^^^^^^^^ help: use `dyn`: `dyn u8 + 'static`
  |
  = note: `#[warn(bare_trait_objects)]` on by default
```
